delocate/delocating.py

```diff
@@ -277,20 +330,33 @@ def delocate_path(tree_path, lib_path,
         is a file in the path depending on ``copied_lib_path``, and the value
         is the ``install_name`` of ``copied_lib_path`` in the depending
         library.
+
+    Raises
+    ------
+    DependencyNotFound
+        When any dependencies can not be located.
     """
     if lib_filt_func == "dylibs-only":
         lib_filt_func = _dylibs_only
+    elif isinstance(lib_filt_func, str):
+        raise TypeError('lib_filt_func string can only be "dylibs-only"')
+    lib_filt_func = lib_filt_func or (lambda _: True)
     if not exists(lib_path):
         os.makedirs(lib_path)
-    lib_dict = tree_libs(tree_path, lib_filt_func)
-    if copy_filt_func is not None:
-        lib_dict = dict((key, value) for key, value in lib_dict.items()
-                        if copy_filt_func(key))
-    copied = delocate_tree_libs(lib_dict, lib_path, tree_path)
-    return copy_recurse(lib_path, copy_filt_func, copied)
+
+    lib_dict = {}  # type: Dict[Text, Dict[Text, Text]]
+    for library_path in walk_directory(tree_path, lib_filt_func):
+        for depending_path, install_name in get_dependencies(library_path):
+            if copy_filt_func and not copy_filt_func(depending_path):
+                continue
+            lib_dict.setdefault(depending_path, {})
```

Comment:
I guess you could do:

```
from collections import defaultdict

    lib_dict = defaultdict(dict)
```

to avoid this line in the loop.

Reply:
I assume I considered this.  I might have worried about how a defaultdict would behave when sent into other functions.

It's a lot to add to remove one line.  I prefer the current version.

Reply:
Fair enough.  I see what you mean.  The fix is just to convert it to a dict on the way out ...  Your call.